src/java.base/share/classes/sun/nio/ch/SelectorImpl.java

```diff
@@ -225,6 +226,8 @@ protected final SelectionKey register(AbstractSelectableChannel ch,
             keys.remove(k);
             k.cancel();
             throw e;
+        } catch (CancelledKeyException ignored) {
```

Comment:
Hi Brian, what about adding the selection key to “keys”  after we set the interesOps ? I do not see any side effect of doing this.

         // it may be observed by a thread iterating over the selector's key set.
-        keys.add(k);
         try {
             k.interestOps(ops);
+            keys.add(k);
         } catch (ClosedSelectorException e) {

Reply:
> Hi Brian, what about adding the selection key to “keys” after we set the interesOps ? I do not see any side effect of doing this.

That would mean a selection key could be selected before it added to the selector's key. So I don't think it is feasible, at least not without working through the spec implications.

Reply:
agreed,as soon as we set the  interestOps on key, native  system call can select the particular fd which is not yet added to the selector key set.

Reply:
One suggestion is to a comment into the empty block, maybe something like "// cancelled by someone iterating over the selector's key set"